Test individual parsing functions

rgreinho opened the pull request:
Types of changes
----------------
<!--
What types of changes does your code introduce?
Select all the choices that apply:
-->
- Code cleanup / Refactoring

Description
-----------
<!--
Describe your changes in detail.
Add a screenshot if applicable.
-->

This patch ensures each individual parsing function is tested
individually instead of using the main parsing entry point.

It also adds more mock data for testing the field parsing mechanisms.

<!--
Motivation and Context
Why is this change required? What problem does it solve? -->


<!--
How Has This Been Tested?
Add any information that could help the reviewer to validate the PR.
Please describe in detail how you tested your changes, include details
of your testing environment, and the tests you ran to see how your
change affects other areas of the code, etc.
-->


Checklist:
----------
<!--
Go over all the following points, and put an `x` in all the boxes that
apply. If you're unsure about any of these, don't hesitate to ask.
We're here to help!
-->

-  [] I have updated the documentation accordingly
-  [] I have written unit tests

<!--
Place the URL of the issue here it this PR fixes an existing issue.
Use either the *FULL* URL (preferred) or the `Username/Repository#`
syntax.
-->

rgreinho commented:
@mscarey It would be cool if you could create another PR to fix the tests I marked as `xfail`. I was not able to fully understand how to use these functions. 🤔

mscarey commented on tests/core/test_parsing.py:
```diff
@@ -234,60 +237,63 @@ def test_parse_notes(page, start, end):
     assert notes.endswith(end)
 
 
-def test_no_DOB_field_when_DOB_not_provided():
```
Why delete this test? Is it covered somewhere else?

rgreinho replied:
Yes, it was a duplicate of this use case: https://github.com/rgreinho/scrapd/blob/remove-test-pages/tests/core/test_person.py#L71-L78

mscarey replied:
I don't think it is covered. The test you kept just checks that the needed fields are correct, not that there's no unneeded field. I needed to add `test_no_DOB_field_when_DOB_not_provided` to catch a bug I was seeing when I ran the CLI.

I'm sure it would be feasible to consolidate them into one test that covered both issues, but it's not there yet.

rgreinho replied:
Ah I see! Good point. That's sort of the negative of the test I pointed you too then. I'll readd it.

Also, with the models we won't need to to that 🎉

rgreinho replied:
Actually we can keep it deleted, as this the next series of test I am working on will test that the output looks exactly how it is expected to look. Therefore the negative tests will be covered for free.

mscarey commented on tests/core/test_parsing.py:
```diff
@@ -234,60 +237,63 @@ def test_parse_notes(page, start, end):
     assert notes.endswith(end)
 
 
-@pytest.mark.parametrize('page,start,end', (
```
Are you deleting 10 tests here and replacing them with only 5? I think the trend should be to increase test coverage. We're going to keep needing to update the parser for new variations on the page formats, and it's very likely that will break the parsing process for some old variations.

rgreinho replied:
When I went through these tests, it looked like some use cases where duplicated, so I was able to knock them down to half. Again, I don't want to test the page parsing here, but the specific parsing functions.

mscarey replied:
Remember when you were surprised to hear me say I thought the parser was brittle...my code for finding the Notes field was part of what I was talking about. How about if you keep the single-function unit tests you created, but you also keep the old tests that didn't get converted to unit tests, at least until there's something to replace them with that serves exactly the same purpose?

rgreinho replied:
We could, but I found a lot of things to redo in our tests, so the full effort will probably cover this. We can formalize it with an issue if you want.

mscarey commented:
> @mscarey It would be cool if you could create another PR to fix the tests I marked as `xfail`. I was not able to fully understand how to use these functions. 🤔

Okay, see #196.

rgreinho commented on tests/core/test_parsing.py:
```diff
@@ -463,52 +469,41 @@ def test_parse_page_01(mocker, filename, expected):
                       ['Ernesto Gonzales Garcia, H/M, (DOB: 11/15/1977)'],
                       id='colon after DOB'))))
-def test_parse_deceased_field_00(input_, expected):
+def test_parse_decease_field_00(input_, expected):
```
Ah yes good catch!

rgreinho replied:
Addressed ✅

rgreinho commented on tests/mock_deceased.py:
```diff
@@ -0,0 +1,58 @@
+Anyone with information regarding this case should call APD’s Vehicular Homicide Unit at (512) 974-4424. You can also submit tips by downloading Austin PD’s free mobile app on <a href="https://austintexas.us5.list-manage.com/track/click?u=1861810ce1dca1a4c1673747c&amp;id=7712ff1eed&amp;e=bcdeacc118">iPhone</a> and <a href="https://austintexas.us5.list-manage.com/track/click?u=1861810ce1dca1a4c167 ...
+"""
+
+DECEASED_00 = """
```
I totally agree with you, but I don't want to test the page parsing here, I want to test the specific function extracting the information from a field which was found. That's why the mocks are so specialized.

mscarey replied:
Does that mean testing that the element-finding functions output the mock data, and then testing that the data extraction functions can get the right text if they start from the mock data? That sounds okay.

rgreinho replied:
Yes, exactly. Only the tests for the 3 high level parsing functions we have should use the pages. The rest should use mocks.

rgreinho commented:
I'll add another test suite to test the page parsing (input: full page, output dict), but we must also be sure that we test the parsing functions individually. The tests really help me (and the reviewers in general) understand the code, and also I often use some test cases when I onboard new contributors.

rgreinho commented:
Here is the test suite branch I want to push on top of this one: https://github.com/rgreinho/scrapd/blob/reorganize-test-suites/tests/core/test_parsing.py. It highlights a lot of problems that we need to tackle before moving forward. That's why I changed my strategy a little and plan to focus on the suite for a bit longer than anticipated.

Bottom line is:
- With the future tests, no need for negative tests for the full parsing
- I'll re-add the notes files tests if you can ensure me we don't test twice the same thing. When I went though the files, it looked to me like we had duplicated use cases, and I removed them. Having duplicate tests does not help with coverage, and provide and false sense of testing.
- The HTML tests won't be eliminated! Far from that! Just reduced to the high level parsing functions: `parsing.parse_twitter_fields`, `parsing.parse_page_content`, and `parsing.parse_page`.
- The parsing fields functions should be tested individually using mocks.

rgreinho commented:
Superseeded by #199.
